README.md
[![CircleCI](https://circleci.com/bb/digitalfoodie/bunbury-storefront/tree/master.svg?style=svg&circle-token=561b699a8a0f958458895300a880c2972a7e9dcd)](https://circleci.com/bb/digitalfoodie/bunbury-storefront/tree/master)

# Storefront

Bunbury Storefront implementation for use with Headless Commerce Platform.

## Technologies

Built using a number of open source projects:

- [Yarn](https://yarnpkg.com/lang/en/) - Dependency management
- [React](https://reactjs.org/) - A JavaScript library for building user interfaces
- [Next.js](https://github.com/vercel/next.js) - React Framework for Server-Rendered Apps.
- [Material UI](https://material-ui.com/) - React Components for Material Design (currently in transition to it)
- [GraphQL](https://graphql.org/) - Providing data for frontend
- [Cypress](https://www.cypress.io/) - End-to-end testing
- [Jest](https://facebook.github.io/jest/) - Unit testing (currently we are trying to achieve 80% test coverage)

## Getting Started

Create `.env` file in **project root** and fill with provided credentials:

```sh
DG_API_KEY=
DG_ORGANIZATION_ID=
DG_API_URL=
DEFAULT_STORE_ID=
LOOKUP_COUNTRY_CODE=au

NOTE: Get the complete list from your internal documentation page
```

```sh
yarn install             # Install dependencies (NodeJS 12.x or higher is recommended )
yarn dev                 # Start development server on port 3000
```

## Running in Production

```sh
yarn build:production    # Builds the application for production usage
yarn start               # Starts a Next.js production server
```

## Recommended Development Practices

- All feature development should happen in scoped feature branches
- Feature branches are created from the up-to-date master branch
- Merging to master branch only via pull requests

Branch naming convention:

```sh
feature/${JIRA_TICKET_ID}-${SHORT_DESCRIPTION}
```

## CI/CD

The project uses [CircleCI](https://app.circleci.com/) for test automation, version increments, building and pushing docker images to docker repository.

Test will be ran for all pushed branches.
Tests are required to pass in order to accept a pull request.

Commits to master branch initiate the following workflow:

1. Tests are run

2. Semantic version is incremented

   - package.json version number is increased
   - matching git tag is created and pushed

3. Manual approval is required for deploying to different environments

4. Once approved, a Docker image is built and pushed to Docker Repository

Docker image naming convention:

```sh
${REGISTRY}/${CLIENT-NAME}-${ENVIRONMENT}:${VERSION_NUMBER}
```

## Optional

### Components documentation

We have examples of our components in `Readme.md` files. You can list a docs with below command:

```sh
npx styleguidist server
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [2.0.15](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v2.0.14...v2.0.15) (2022-03-31)

### [2.0.14](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v1.1.85...v2.0.14) (2022-03-31)

### [2.0.13](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v1.1.84...v2.0.13) (2022-03-29)

### [2.0.12](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v2.0.11...v2.0.12) (2022-03-24)

### [2.0.11](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v1.1.80...v2.0.11) (2022-03-24)

### [2.0.10](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v2.0.9...v2.0.10) (2022-03-22)

### [2.0.9](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v2.0.8...v2.0.9) (2022-03-22)

### [2.0.8](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v1.1.77...v2.0.8) (2022-03-22)

### [2.0.7](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v2.0.6...v2.0.7) (2022-03-18)

### [2.0.6](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v1.1.75...v2.0.6) (2022-03-18)

### [2.0.5](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v2.0.4...v2.0.5) (2022-03-17)

### [2.0.4](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v1.1.74...v2.0.4) (2022-03-17)

### [2.0.3](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v2.0.2...v2.0.3) (2022-03-16)

### [2.0.2](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v1.1.73...v2.0.2) (2022-03-15)

### [1.1.29](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v1.1.28...v1.1.29) (2022-03-10)

### [1.1.28](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v1.1.27...v1.1.28) (2022-03-09)

### [1.1.27](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v1.1.70...v1.1.27) (2022-03-09)

### [1.1.26](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v1.1.25...v1.1.26) (2022-03-09)

### [1.1.25](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v1.1.24...v1.1.25) (2022-03-09)

### [1.1.24](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v1.1.23...v1.1.24) (2022-03-09)

### [1.1.23](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v1.1.22...v1.1.23) (2022-03-09)

### [1.1.22](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v1.1.69...v1.1.22) (2022-03-09)

### [1.1.21](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v1.1.66...v1.1.21) (2022-03-07)

### [1.1.20](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v1.1.57...v1.1.20) (2022-02-22)

### [1.1.19](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v1.1.18...v1.1.19) (2022-02-22)

### [1.1.18](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v1.1.17...v1.1.18) (2022-02-22)

### [1.1.17](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v1.1.16...v1.1.17) (2022-02-22)

### [1.1.16](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v1.1.15...v1.1.16) (2022-02-21)

### [1.1.15](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v1.1.14...v1.1.15) (2022-02-21)

### [1.1.14](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v1.1.56...v1.1.14) (2022-02-21)

### [1.1.13](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v1.1.51...v1.1.13) (2022-02-15)

### [1.1.12](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v1.1.11...v1.1.12) (2022-01-17)

### [1.1.11](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v1.1.10...v1.1.11) (2022-01-17)

### [1.1.10](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v1.1.9...v1.1.10) (2022-01-13)

### [1.1.9](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v1.1.8...v1.1.9) (2022-01-13)

### [1.1.8](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v1.1.7...v1.1.8) (2022-01-13)

### [1.1.7](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v1.1.6...v1.1.7) (2022-01-13)

### [2.0.13](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v2.0.12...v2.0.13) (2022-03-29)

### [2.0.12](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v2.0.11...v2.0.12) (2022-03-24)

### [2.0.11](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v2.0.10...v2.0.11) (2022-03-24)

### [2.0.10](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v2.0.9...v2.0.10) (2022-03-22)

### [2.0.9](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v2.0.8...v2.0.9) (2022-03-22)

### [2.0.8](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v2.0.7...v2.0.8) (2022-03-22)

### [2.0.7](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v2.0.6...v2.0.7) (2022-03-18)

### [2.0.6](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v1.1.75...v2.0.6) (2022-03-18)

### [2.0.5](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v2.0.4...v2.0.5) (2022-03-17)

### [2.0.4](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v1.1.74...v2.0.4) (2022-03-17)

### [2.0.3](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v2.0.2...v2.0.3) (2022-03-16)

### [2.0.2](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v1.1.73...v2.0.2) (2022-03-15)

### [1.1.29](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v1.1.28...v1.1.29) (2022-03-10)

### [1.1.28](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v1.1.27...v1.1.28) (2022-03-09)

### [1.1.27](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v1.1.70...v1.1.27) (2022-03-09)

### [1.1.26](https://bitbucket.org/digitalfoodie/bunbury-storefront/compare/v1.1.25...v1.1.26) (2022-03-09)


config.js
/**
 * All config needed for the app should be parsed from environment variables in this file.
 * This goes for config needed in server code as well as config needed in browser code.
 * - If the config is needed in a Node js file, you should import it directly from here.
 * - If the config is needed in browser code, which may sometimes be executed in a browser
 *   and sometimes be executed on the server, you should import the config from here into
 *   next.config.js and add it to either `publicRuntimeConfig` or `serverRuntimeConfig`.
 *   Then use `getConfig()` to get it within React components and other browser code.
 */
const envalid = require('envalid')

// Config is loaded for the `next build` command, too, and we don't want to complain
// about missing environment variables in that phase.
if (process.env.IS_BUILDING_NEXTJS) {
  module.exports = {}
} else {
  const { port, str, url, num, bool } = envalid

  /**
   * See https://www.npmjs.com/package/envalid
   *
   * Envalid parses NODE_ENV automatically, and provides the following
   * shortcut (boolean) properties for checking its value:
   *   env.isProduction    // true if NODE_ENV === 'production'
   *   env.isTest          // true if NODE_ENV === 'test'
   *   env.isDevelopment   // true if NODE_ENV === 'development'
   *
   * Be sure to add `{ default: null }` if it should be optional.
   */
  module.exports = envalid.cleanEnv(
    process.env,
    {
      DG_API_KEY: str(),
      DG_ORGANIZATION_ID: str(),
      DG_API_URL: str(),
      SHOPPER_API_URL: str({ default: '' }),
      DEFAULT_STORE_ID: str(),
      SERVER_URL: url({ default: 'https://bfm-test.vercel.app' }),
      NODE_ENV: str({ choices: ['development', 'test', 'jesttest', 'production'], default: 'production' }),
      PORT: port({ default: 3000 }),
      LOOKUP_COUNTRY_CODE: str({ default: 'au' }),
      CLUSTER: str({ default: 'dev' }),
      BUILD_DATE: str({ default: new Date().toLocaleString() }),
      BUILD_NUMBER: str({ default: 'localhost' }),
      BUILD_IMAGE: str({ default: 'web-client-localhost' }),
      GOOGLE_TAG_MANAGER_KEY: str({ default: 'GTM-565D52D' }),
      GOOGLE_MAP_KEY: str({ default: 'AIzaSyClAuZvW6KiuNiDNktdHeCGS7GAg4puMXg' }), // Never add a real customer map key to default
      FACEBOOK_DOMAIN_VERIFICATION: str({ default: '' }),
      BRAND: str({ default: "Bunbury Farmer's Market" }),
      BRAND_THEME: str({ default: 'false' }),
      SERVICE_WORKER_TIMEOUT: num({ default: 60 * 1000 }),
      SITE_URL: str({ default: 'https://bfm-test.vercel.app' }),

      NEXTAUTH_URL: str({ default: 'https://bfm-test.vercel.app' }),
      AUTH_CLIENT_ID: str({ default: '' }),
      AUTH_CLIENT_SECRET: str({ default: '' }),
      AUTH_TENANT_NAME: str({ default: '' }),
      AUTH_TENANT_ID: str({ default: '' }),
      AUTH_LOGIN_FLOW: str({ default: '' }),
      AUTH_PROFILE_EDIT_URL: str({ default: '' }),
      AUTH_PASSWORD_RESET_URL: str({ default: '' }),
      AUTH_SESSION_MAX_AGE: num({ default: 30 * 24 * 60 * 60 }), // 30 days
    },
    {
      // disable dotenv processing
      dotEnvPath: '.env',
      // https://www.npmjs.com/package/envalid#strict-mode
      strict: false,
    },
  )
}
